diff-tests/core/model/src/model.ts

```diff
@@ -0,0 +1,671 @@
+  m;
+  // height of onChanOpenConfirm event
+  initialHeight: number;
+  // next id to use
+  vscID: number;
+  // map ids to height of sending
+  // used to calculate infraction height from consumer initiated slashing
+  vscIDtoH: Record<number, number>;
+  // map ids to unbonding operation ids
+  // used to mature unbonding operations when receiving maturity packets
+  vscIDtoOpIDs: Map<number, number[]>;
+  // validators who have been slashed since last VSC
+  downtimeSlashAcks: Validator[];
+  // is a validator tombstoned?
+  tombstoned: boolean[];
+  // unbonding operations to be completed in EndBlock
+  matureUnbondingOps: number[];
+
+  constructor(model: Model, { ccvP }: ModelInitState) {
+    this.m = model;
+    Object.assign(this, ccvP);
+  }
+
+  endBlock = () => {
+    this.vscIDtoH[this.vscID] = this.m.h[P] + 1;
+    this.matureUnbondingOps.forEach((opID) => {
+      this.m.staking.unbondingCanComplete(opID);
+    });
+    this.matureUnbondingOps = [];
+    const valUpdates = this.m.staking.valUpdates();
+    const hasOps =
+      this.vscIDtoOpIDs.has(this.vscID) &&
+      0 < this.vscIDtoOpIDs.get(this.vscID)!.length;
+    if (0 < _.keys(valUpdates).length || hasOps) {
+      if (0 === _.keys(valUpdates).length) {
+        this.m.events.push(Event.SEND_VSC_NOT_BECAUSE_CHANGE);
+      }
+      if (0 < this.downtimeSlashAcks.length) {
+        this.m.events.push(Event.SEND_VSC_WITH_DOWNTIME_ACK);
+      } else {
+        this.m.events.push(Event.SEND_VSC_WITHOUT_DOWNTIME_ACK);
+      }
+      const data: Vsc = {
+        vscID: this.vscID,
+        updates: valUpdates,
+        downtimeSlashAcks: this.downtimeSlashAcks,
+      };
+      this.downtimeSlashAcks = [];
+      this.m.outbox[P].add(data);
+    }
+    this.vscID += 1;
+  };
+
+  onReceive = (data: PacketData) => {
+    // It's sufficient to use isDowntime field as differentiator
+    if ('isDowntime' in data) {
+      this.onReceiveSlash(data);
+    } else {
+      this.onReceiveVSCMatured(data);
+    }
+  };
+
+  onReceiveVSCMatured = (data: VscMatured) => {
+    if (this.vscIDtoOpIDs.has(data.vscID)) {
+      this.vscIDtoOpIDs.get(data.vscID)!.forEach((opID: number) => {
+        this.matureUnbondingOps.push(opID);
+      });
+      this.vscIDtoOpIDs.delete(data.vscID);
+    }
+  };
+
+  onReceiveSlash = (data: Slash) => {
+    let infractionHeight = undefined;
+
+    if (data.vscID === 0) {
+      infractionHeight = this.initialHeight;
+    } else {
+      infractionHeight = this.vscIDtoH[data.vscID];
+    }
+
+    if (this.m.staking.status[data.val] === Status.UNBONDED) {
+      return;
+    }
+
+    if (data.isDowntime) {
+      this.m.events.push(Event.RECEIVE_DOWNTIME_SLASH_REQUEST);
+    } else {
+      this.m.events.push(Event.RECEIVE_DOUBLE_SIGN_SLASH_REQUEST);
+    }
+
+    if (this.tombstoned[data.val]) {
+      return;
+    }
+
+    this.m.staking.slash(data.val, infractionHeight);
+    this.m.staking.jailUntil(data.val, this.m.t[P] + JAIL_SECONDS);
```

Comment:
A slash validator will be jailed forever. This is especially true since Unjail is not implemented. Until we fix this, we should increase the number of validators for the generated traces (currently `NUM_VALIDATORS = 4`).

Reply:
> Until we fix this, we should increase the number of validators for the generated traces

Why?

Reply:
Any generated trace will have maximum 4 slash events. Moreover, maximum 2 slash events can happen at the same time. Seems little.

Reply:
Hmm, what about keeping it as is for now and making a new issue to track this concern, and working on it in conjunction with

- https://github.com/cosmos/interchain-security/issues/319